Document how closure capturing interacts with discriminant reads

meithecatte opened the pull request:
This is the behavior after the bugfixes in rust-lang/rust#138961. I have successfully ran `mdbook test` with `RUSTUP_TOOLCHAIN` pointed at a stage1 built on top of the aforementioned PR – until it's merged, the CI here will fail.

meithecatte committed:
Document how closure capturing interacts with discriminant reads

This is the behavior after the bugfixes in rustc PR 138961.

meithecatte committed:
Document how range and slice patterns can constitute discriminant reads

Nadrieril commented on src/types/closure.md:
```diff
@@ -264,6 +267,69 @@ let c = || {
+r[type.closure.capture.precision.discriminants]
+### Capturing for discriminant reads
+
+If pattern matching requires inspecting a discriminant, the relevant place will get captured by `ImmBorrow`.
```
Unless we mutably access the contents, right? E.g.
```rust
let c = || match x {
    (Example::A(ref mut _y), _) => println!("variant A"),
    (Example::B(_), _) => println!("variant B"),
};
```

Nadrieril replied:
ok yep, that's the "Shared prefix" thing

meithecatte replied:
Yup. Do you think this deserves a clarification? (if so, how would you word it)

Nadrieril reviewed (approved):
The changes look accurate to me. I don't know about things related to how the reference is organized so I'll leave to someone who knows to approve this.

ehuss commented on src/types/closure.md:
````diff
@@ -264,6 +267,118 @@ let c = || {
+```
+
+r[type.closure.capture.precision.discriminants.range-patterns]
+Matching against a [range pattern][patterns.range] constitutes a discriminant read, even if
+the range matches all possible values.
````
I'm not quite following this terminology, particularly around the user of the term "discriminant". My understanding is that a discriminant is specifically a tag associated with an enum. I wouldn't consider an integer to have a discriminant. Can you clarify what this means?

meithecatte replied:
Right, this happened because the relevant rustc has a distinction between (a) binding sub-values to identifiers with various binding modes, and (b) discriminant reads—with other "shallow/immediate" reads, such as the length field of a slice, being shoe-horned into the category of discriminant reads.

From the perspective of reference text, this obviously doesn't make sense. I'll rephrase that as just "constitutes a read", I think, unless you have any better ideas on how to word this?

ehuss commented on src/types/closure.md:
````diff
@@ -264,6 +267,118 @@ let c = || {
+```
+
+r[type.closure.capture.precision.discriminants.slice-patterns]
+Matching against a [slice pattern][patterns.slice] constitutes a discriminant read if
+the slice pattern needs to inspect the length of the scrutinee.
````
Same question about the use of the term "discriminant".

ehuss reviewed (commented):
Thanks, can you also unwrap any new text being added?

meithecatte committed:
Don't call things "discriminant reads" just because they behave like ones

meithecatte committed:
Clarify what gets read by slice patterns

meithecatte commented on src/types/closure.md:
```diff
@@ -344,11 +340,10 @@ c();
 r[type.closure.capture.precision.discriminants.slice-patterns]
-Matching against a [slice pattern][patterns.slice] constitutes a discriminant read if
-the slice pattern needs to inspect the length of the scrutinee.
+Matching against a [slice pattern][patterns.slice] performs a read if the slice pattern needs to inspect the length of the scrutinee. The read will cause the closure to borrow the relevant place by `ImmBorrow`.
```
Okay, how about this new revision?

Nadrieril commented:
This looks good for me, could someone with authority approve this plz?

traviscross committed:
Revise text about capture precision

To the language about a capture path, this branch adds, "..., as well
as any further projections performed by matching against patterns",
and it links to the section on capturing and wildcard patterns.

The link is a bit unobvious, since the section discusses what's not
captured.  The "as well as" phrasing is a bit awkward, and it's
introduced after a comma.  That indicates a non-restrictive clause,
which this is not.

The next rule talks about what place projections are.  It seems that
it might be better to instead extend that language than to try to fit
this in here.  We can say that destructuring can effect a place
projection, and that seems right.  Let's do that.

traviscross committed:
Revise intro on wildcard patterns and capturing

Let's tighten up the wording a bit in the introduction to wildcard
patterns and capturing to make it better match how we'll present
things below.  We'll remove the "for example..." ahead of the example,
as we're not doing this below, and we'll separately extend the
comments within the examples to include the relevant information.

traviscross committed:
Add example of what's not captured w.r.t. destructuring

In this branch, we're adding a statement in the wildcard section about
how destructuring tuples, structs, and single-variant enums, by
itself, does not constitute a read.  Let's add some examples of this
to demonstrate.

traviscross committed:
Make statement more self-standing

Rather than saying "this also includes", let's make this statement
stand better on its own.  This will be helpful when giving it a rule
identifier, and given how this statement is some distance, visually,
from the other statements due to the examples, it adds clarity.

traviscross committed:
Add rule identifier to destructuring/capturing claim

Let's give the claim about the effect of destructuring with respect to
reads and capturing a rule identifier.

traviscross committed:
Fix `non_exhaustive` rule identifier fragment

When a fragment of a rule identifier matches a Rust identifier, we use
an underscore rather than a dash as a separator.  Let's do that here.

traviscross committed:
Add forward reference for `non_exhaustive` exception

The normative rule about destructuring and capturing does not itself
state the exception for `non_exhaustive` enums.  That exception
follows below.  Let's add an admonition to link down to it so it's not
missed.

It's not great having this duplication; it'd be better if we weren't
restating this fact twice.  But it's not clear what to do about this.
The first statement is making a broader statement about tuples,
structs, and single-variant enums.  The section that follows below is
focused on discriminants.  We'll think about this more later.

traviscross committed:
Revise text about field matching and `..`

Let's clean up the wording of this claim a bit.  Let's put capturing
in the present tense, rather than in the future tense, as that's how
we're going to continue below.  Let's take out, as well, the explicit
lead-in to the example, as we're not generally doing that in these
sections.

traviscross committed:
Add identifier for wildcard fields capturing rule

Let's add a rule identifier for the rule about wildcards and the
capturing of fields.

traviscross committed:
Remove example lead-in from array slice capturing rule

Let's remove the explicit lead-in to the example from the rule about
(the lack of) partial captures of arrays and slices.

traviscross committed:
Remove redundant example on structs and capturing

What this example is showing is how one field of a struct can be
captured while another is not (when using rest patterns).  However,
the example above demonstrates this with tuples, and there are no
interesting differences between tuple fields and struct fields in this
case.

traviscross committed:
Improve examples in wildcard capturing section

Some of the examples in this section did not really prove the claim
they were making; let's adjust them to do this.  Let's also simplify
and minimize them to match the style of other recent examples we're
adding and that of the ones that will follow in the section below.

One particular stylistic conundrum is what to do about comments like
this:

    // ERROR: Borrow of moved value.

Normally, our stylistic convention would be to capitalize after the
comma in this case and end with a period.  That's what we'd do, e.g.,
in a similar case like this:

    // OK: The value can be moved here.

But, of course, `rustc` doesn't capitalize and add a period to these
kind of error messages, making it tempting to follow that lead.

Since we already don't always use the same error messages that `rustc`
does -- it's not a goal to match those -- it seems better to be
internally consistent with our own documentation norms.  Let's
capitalize and add the period.  We'll later add this to the style
guide and work to align the document with this.

traviscross committed:
Clarify main discriminant reads rule

The rule,

> If pattern matching requires inspecting a discriminant, the relevant
> place will get captured by `ImmBorrow`.

is unclear in two ways.  For one, when does pattern matching "require
inspecting" the discriminant?  Second, what is the "relevant" place?

Let's break this up into two rules (and add rule identifiers).  The
first rule will state that reading the discriminant captures the place
containing the discriminant (by `ImmBorrow`).  The second rule will
state explicitly the baseline rule for when such reads happen.

traviscross committed:
Revise examples for main discriminant reads rule

The example here showed what isn't captured, but it didn't demonstrate
what is captured.  Let's show both.  We'll specifically use a
non-`Copy` type for this because [Rust PR #138961] affects the
behavior in the non-`Copy` case.

[Rust PR #138961]: https://github.com/rust-lang/rust/pull/138961

traviscross committed:
Revise `...discriminants.single-variant` rule

Let's simplify the wording of the rule a bit, reiterate the effect of
the rule (as we're doing elsewhere), and tighten up the example.

traviscross committed:
Revise `...discriminants.non_exhaustive` rule

Let's clarify some wording in this rule, making it a bit more
self-standing, and inline the link target as, when using the rule
identifier for this, it's not meaningfully longer than doing it the
other way.

traviscross committed:
Revise `...discriminants.uninhabited-variants` rule

Let's rename this rule identifier to the plural; we generally use the
plural whenever it can make sense, and it does here.  Let's make the
wording of the rule more clear and self-standing.  And let's tighten
up the example.

traviscross committed:
Revise `...discriminants.range-patterns` rule

Let's tighten up the text for this rule and condense it down into one
sentence.  It's a bit awkward to have another sentence just to say,
"this is true even if...".  Let's clean up the example as well.

traviscross committed:
Revise `...discriminants.slice-patterns-*` rules

The section here stated:

> Matching against a slice pattern that needs to inspect the length of
> the scrutinee performs a read of the pointer value in order to fetch
> the length. The read will cause the closure to borrow the relevant
> place by `ImmBorrow`.

It then goes on to state exceptions for arrays matched against slice
patterns and slice patterns containing only a rest pattern.

In saying that the pointer value is read, this might suggest that the
pointer is captured.  Actually, though, it's the slice (the pointee)
that is captured.  Let's fix this.

Beyond that, as we saw in an earlier commit, it's better to not lean
on the reader to infer the "relevant place" or when the length needs
to be inspected.  Let's elaborate those details and state the full
rule in one go, upfront, and then state a separate guarantee that
matching an array against a slice pattern does not do a read.

We'll also fix a typo, add rule identifiers, and tighten up the
examples.

traviscross committed:
Separate sections for range/slice pattern capturing

The rules for the capturing behavior related to range patterns and
slice patterns were comingled with the rules for capturing and
discriminant reads.  While these are spiritually related, they are
distinct, so let's break these apart into separate sections.

meithecatte commented on src/types/closure.md:
````diff
@@ -264,6 +266,114 @@ let c = || {
+
+```rust,compile_fail,E0506
+let x: &mut [i32] = &mut [1, 2, 3];
+let c = || match x { // captures `*x` by ImmBorrow
````
I don't have the bandwidth to test this myself right now, but I assume you're hitting this optimization: https://github.com/rust-lang/rust/blob/016baacaf6ed5698625036e3afe3900535015680/compiler/rustc_hir_typeck/src/upvar.rs#L2543-L2572

My understanding, based on knowledge that has faded a bit already, is that the language semantics says we capture `*x`, but the actual compiled output optimizes that to capturing `x` instead, on account of that not being observable. I assume that if you stare 5–10 minutes at the doc comment I linked, you'll be able to come up with an example where the optimization doesn't apply and the capture of `*x` can be observed in the MIR.

traviscross committed:
Add admonition about the desugaring of destructuring

For our purposes in describing capture paths, we define place
projections.  For the text that follows in the sections below to hold,
we need destructuring to be treated as a place projection for this
purpose.

However, it might seem surprising for destructuring to be in this
list.  While it is believed to have the same effect as a projection
expression, we might not consider an instance of pattern destructuring
to be a projection expression exactly.

To better contextualize this, let's add an admonition that mentions
that pattern destructuring desugars into the kind of field accesses
that we would more likely think about as projection operations.

traviscross committed:
Add admonition about pointer vs pointee capturing

There's some surprising nuance to what gets captured when reading the
length of a slice with a slice pattern.  Let's add an admonition about
this.

traviscross committed:
Fix preexisting typo on "ancestors"

While we're here, let's fix a typo on the word "ancestors" in this
section.

Nadrieril commented:
EDIT: nevermind, moved this to another issue

meithecatte commented on src/types/closure.md:
````diff
@@ -185,85 +189,259 @@ If this were to capture `m`, then the closure would no longer outlive `'static`,
+```
+
+r[type.closure.capture.precision.discriminants.non_exhaustive]
+If [`#[non_exhaustive]`][attributes.type-system.non_exhaustive] is applied to an enum defined in an external crate, the enum is treated as having multiple variants for the purpose of deciding whether a read occurs, even if it actually has only one variant.
````
> Can removing `#[non_exhaustive]` cause code in another crate to stop compiling?

Unfortunately, yes. If we have the following upstream library `crate1`:

```rust
#[non_exhaustive]
pub enum Meow {
    A(u32, String),
}
```

and a downstream crate with the following code

```rust
use crate1::Meow;

pub fn f(x: Meow) -> impl FnOnce() {
    || {
        match x {
            Meow::A(a, b) => {
                drop((a, b));
            }
            _ => unreachable!(),
        }
    }
}
```

then with rust-lang/rust#138961, removing `#[non_exhaustive]` breaks the downstream crate. This would be another argument for the change proposed in rust-lang/rust#147722, i.e. that regardless of `#[non_exhaustive]`, there should be a read of the discriminant.

theemathas replied:
> Can adding/removing `#[non_exhaustive]` affect run time behavior in closure captures?

Turns out the answer is **yes**. Consider the following code:

dep/src/lib.rs:
```rust
#[non_exhaustive]
pub enum Thing {
    One(i32, LoudDrop),
}

pub struct LoudDrop(pub &'static str);
impl Drop for LoudDrop {
    fn drop(&mut self) {
        println!("dropping {}", self.0);
    }
}
```

src/main.rs:
```rust
use dep::{LoudDrop, Thing};

#[allow(unused)]
fn main() {
    let mut thing = Thing::One(0, LoudDrop("a"));
    let closure = move || match thing {
        Thing::One(x, _) => {}
        _ => unreachable!(),
    };
    println!("before assign");
    thing = Thing::One(1, LoudDrop("b"));
    println!("after assign");
}
```

Using nightly rust, this code prints:
```
before assign
after assign
dropping a
dropping b
```

However, if we delete the `#[non_exhaustive]`, this code prints:
```
before assign
dropping a
after assign
dropping b
```

It seems that the `LoudDrop` is moved into the closure only if the enum is `#[non_exhaustive]`

meithecatte replied:
Ah, I was considering something like this but couldn't get it to a full example because I forgot about drop being observable :sweat_smile:

ehuss reviewed (approved):
Thanks! Merging now that https://github.com/rust-lang/rust/pull/138961 has merged.

RalfJung commented on src/types/closure.md:
````diff
@@ -185,85 +189,259 @@ If this were to capture `m`, then the closure would no longer outlive `'static`,
+```
+
+r[type.closure.capture.precision.wildcard.destructuring]
+Destructuring tuples, structs, and single-variant enums does not, by itself, cause a read or the place to be captured.
````
It is "the same" in the sense that it's all caused by special-casing single-variant enums except when they are foreign-crate non-exhaustive enums, isn't it?

theemathas replied:
Fair enough. They're "the same" in that sense.
